server/src/main/java/org/apache/druid/metadata/IndexerSQLMetadataStorageCoordinator.java

```diff
@@ -1886,7 +1856,67 @@ private SegmentIdWithShardSpec createNewSegment(
               committedMaxId == null ? 0 : committedMaxId.getShardSpec().getNumCorePartitions()
           )
       );
+      return getTrueAllocatedId(allocatedId);
+    }
+  }
+
+  private SegmentIdWithShardSpec getTrueAllocatedId(
+      SegmentIdWithShardSpec allocatedId
+  )
+  {
+    // Check if there is a conflict with an existing entry in the segments table
+    if (retrieveSegmentForId(allocatedId.asSegmentId().toString(), true) == null) {
+      return allocatedId;
+    }
+
+    // If yes, try to compute allocated partition num using the max unused segment shard spec
+    SegmentIdWithShardSpec unusedMaxId = getUnusedMaxId(
+        allocatedId.getDataSource(),
+        allocatedId.getInterval(),
+        allocatedId.getVersion()
+    );
+    // No unused segment. Just return the allocated id
+    if (unusedMaxId == null) {
+      return allocatedId;
+    }
+
+    int maxPartitionNum = Math.max(
+        allocatedId.getShardSpec().getPartitionNum(),
+        unusedMaxId.getShardSpec().getPartitionNum() + 1
```

Comment:
Rather than doing unusedMaxId + 1, we should probably not use this version at all, if there is any existing unused segment for that version.

I think it would be better to filter out all the invalid versions:
- used but overshadowed versions (fixed in #15459 )
- unused versions

Reply:
Isn't it possible that someone dropped the last segment and there's a conflict because of that one segment?